pkg/proc/target_group.go

```diff
@@ -0,0 +1,125 @@
+package proc
+
+import (
+	"fmt"
+	"strings"
+)
+
+// TargetGroup reperesents a group of target processes being debugged that
+// will be resumed and stopped simultaneously.
+// New targets are automatically added to the group if exec catching is
+// enabled and the backend supports it, otherwise the group will always
+// contain a single target process.
+type TargetGroup struct {
+	targets []*Target
+	procs   []ProcessInternal
+	Sel     *Target
+
+	RecordingManipulation
+	recman RecordingManipulationInternal
+
+	// KeepSteppingBreakpoints determines whether certain stop reasons (e.g. manual halts)
+	// will keep the stepping breakpoints instead of clearing them.
+	KeepSteppingBreakpoints KeepSteppingBreakpoints
+
+	LogicalBreakpoints map[int]*LogicalBreakpoint
+
+	continueOnce ContinueOnceFunc
+	cctx         *ContinueOnceContext
+}
+
+// NewGroup creates a TargetGroup containing the specified Target.
+func NewGroup(t *Target) *TargetGroup {
+	if t.partOfGroup {
+		panic("internal error: target is already part of a group")
+	}
+	t.partOfGroup = true
+	return &TargetGroup{
+		RecordingManipulation: t.recman,
+		targets:               []*Target{t},
+		procs:                 []ProcessInternal{t.proc},
+		Sel:                   t,
+		cctx:                  &ContinueOnceContext{},
+		recman:                t.recman,
+		LogicalBreakpoints:    t.Breakpoints().Logical,
+		continueOnce:          t.continueOnce,
+	}
+}
+
+// Targets returns a slice of targets in the group.
+func (grp *TargetGroup) Targets() []*Target {
+	return grp.targets
+}
+
+// Valid returns true if any target in the target group is valid.
+func (grp *TargetGroup) Valid() (bool, error) {
```

Comment:
Is this the correct behavior? Since the `TargetGroup` operates on every `Target` within it, wouldn't we want `Valid` to be true if *all* targets are valid?

Reply:
As long as one of the processes is valid (= not quit) the group is valid. Child processes can die and leave the group.